Add workflow documentation (#40)
exercism/development-environment: docs/common_workflows.md

# Getting the development environment set up

## What your directory should look like

```text
<root>
├── development-environment
├── <Any other v3 repos you may be working with>
└── v3-website
```

## Git workflow to working from a fork

### Cloning `exercism/v3-website` and maintaining your `fork`

```text
> cd <root>
> git clone https://github.com/exercism/v3-website.git
> cd v3-website
```

> After a pull request is open against `exercism/v3-website`, the remote fork's branch can be checked out [following this guide][github-mod-inactive-pr-local].  After making a commit, you can't `git push` to the `exercism`'s repo, but you can to your own fork

Reviewer: ```suggestion
After a pull request is opened against `exercism/v3-website`, the remote fork's branch can be checked out [following this guide][github-mod-inactive-pr-local].  After making a commit, you can't `git push` to the `exercism`'s repo, but you can to your own fork.
```

```text
> git push <remote-name> HEAD:<remote-branch-name>
```

### Using your own fork

```text
> cd <root>
> git clone <your v3-website fork> v3-website 
```

> Then when making commits you can easily push the commit to the origin

Reviewer: ```suggestion
Then when making commits you can easily push the commit to the origin (your fork).
```

## Running tests

Now when you start the development environment, it will use your `v3-website` repo as the source for the rails application. When the development environment is running, enter the website's running docker container to run tests.

```text

> > docker exec -it development-environment_website_1 bash

Reviewer: There is an easier (and more foolproof) way:

```suggestion
> ./bin/shell website
```

Author: I like it! Maybe I would first use the short way you suggest, then explain what that does so people can understand the magic.

Author: ```suggestion
> ./bin/shell website

# This is a short cut for: 
# > docker exec -it development-environment_website_1 bash
```

> bundle exec rails test # run all rails tests
> yarn test [<specific test file>] # run jest tests
```

> 

Reviewer: There are also script versions that one could use instead of the above approach:

```
> ./bin/script website run-js-tests
> ./bin/script website run-tests
> ./bin/script website run-system-tests
```

They don't keep an open connection, but might be easier to remember.

Author: ````suggestion

If you don't need a persistent terminal in the container, you can also use these commands:

```text
> ./bin/script website run-js-tests         # Run the jest tests
> ./bin/script website run-tests            # Run the Rails tests
> ./bin/script website run-system-tests     # Run the Capybara tests
```
````

# Getting the development environment set up

## What your directory should look like

```text
<root>
├── development-environment
├── <Any other v3 repos you may be working with>
└── v3-website
```

## Git workflow to working from a fork

### Cloning `exercism/v3-website` and maintaining your `fork`

```text
> cd <root>
> git clone https://github.com/exercism/v3-website.git
> cd v3-website
```

After a pull request is opened against `exercism/v3-website`, the remote fork's branch can be checked out [following this guide][github-mod-inactive-pr-local].  After making a commit, you can't `git push` to the `exercism`'s repo, but you can to your own fork.

```text
> git push <remote-name> HEAD:<remote-branch-name>
```

### Using your own fork

```text
> cd <root>
> git clone <your v3-website fork> v3-website 
```

Then when making commits you can easily push the commit to the origin (your fork).

## Running tests

Now when you start the development environment, it will use your `v3-website` repo as the source for the rails application. When the development environment is running, enter the website's running docker container to run tests.

```text

> > ./bin/shell website
> 
> # This is a short cut for: 
> # > docker exec -it development-environment_website_1 bash

Author: ```suggestion
# Open a terminal to the running website container:
> ./bin/shell website

# You can alternatively use:
> docker exec -it development-environment_website_1 bash

```

> > bundle exec rails test # run all rails tests

Author: ```suggestion
#

> bundle exec rails test # run all rails tests
```
